assets/src/store.rs

```diff
@@ -0,0 +1,95 @@
+use std::error::Error;
+use std::fs::File;
+use std::io::Error as IoError;
+use std::path::PathBuf;
+use std::sync::atomic::{AtomicUsize, Ordering};
+
+/// An `Allocator`, holding a counter for producing unique ideas for the stores.
+#[derive(Debug, Default)]
+pub struct Allocator {
+    store_count: AtomicUsize,
+}
+
+impl Allocator {
+    /// Creates a new `Allocator`.
+    pub fn new() -> Self {
+        Default::default()
+    }
+
+    /// Produces a new store id.
+    pub fn next_store_id(&self) -> StoreId {
+        StoreId(self.store_count.fetch_add(1, Ordering::Relaxed))
+    }
+}
+
+/// Directory store.
+///
+/// Please note that there is a default directory storage
+/// inside the `Loader`, which is automatically used when you call
+/// `load`. In case you want another, second, directory for assets,
+/// you can instantiate one yourself, too. Please use `Loader::load_from`
+/// then.
+#[derive(Debug)]
+pub struct Directory {
+    id: StoreId,
+    loc: PathBuf,
+}
+
+impl Directory {
+    /// Creates a new directory storage.
+    pub fn new(alloc: &Allocator, loc: PathBuf) -> Self {
+        Directory {
+            id: alloc.next_store_id(),
+            loc: loc,
+        }
+    }
+}
+
+impl Store for Directory {
+    type Error = IoError;
+    type Location = PathBuf;
+
+    fn store_id(&self) -> StoreId {
+        self.id
+    }
+
+    fn load(&self, category: &str, name: &str, ext: &str) -> Result<Vec<u8>, IoError> {
```

Comment:
I completely agree with you about flatter hierarchies. But keep in mind that users can sometimes be astonishingly dumb, haha. Be sure to code defensively and consider all possibilities!